### ISSUE TITLE
Need a way to only update associations.

### ISSUE THREAD

**Issue by lwansbrough** [balderdashy/sails#2297]
Hey guys, so I've run into a blocking problem with the current associations implementation - specifically I'm referring to how we're supposed to update models to attach associations.

Currently, I can do the following:

    User.update({ id: 1 }, { id: 1, password: '123456' }, ...)

Then I can use the `beforeUpdate` lifecycle callback to check `if(values.password)` and hash the new password for storage if it exists in the values to be updated. This works as expected.

A problem exists however, with the way models are updated when an association is changed. Let's say I have a user...

    User.findOne(1, function(err, user) {
      user.roles.add(1);
      user.save(cb);
    });

Should work no problem right? Not so much. Because we haven't explicitly told Sails exactly which values we're updating, it assumes we're updating everything. The result is that `values` in my `beforeUpdate` lifecycle callback now has `values.password` = the database entry for the user's password. My code then assumes that the password is being changed and rehashes the password.

I don't think the problem here is with my usage of the `beforeUpdate` lifecycle callback - I think my use case is pretty typical. The problem is that Sails is assuming I'm updating something which I am not.

Is there a workaround which doesn't require me moving the logic out of the lifecycle callback?

**Comment by devinivy** [comment]
I'm curious what pattern is considered kosher for maintaining a hashed password– what did you settle on, @lwansbrough?

**Comment by lwansbrough** [comment]
@devinivy bcrypt is the most common library - it'll store your password hash and salt together for you.

**Comment by devinivy** [comment]
I meant which waterline pattern.  Currently I'm using `beforeValidate` and only updating the hash if the set value is plaintext.

**Comment by lwansbrough** [comment]
Ah, I added a class method `User.hashPassword` and then an instance method `changePassword` and added a call to `User.hashPassword` in `beforeCreate` then any other calls for password changes will be handled explicitly by a controller.

**Comment by sailsbot** [comment]
Thanks for posting, @lwansbrough.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
